src/main/java/com/amazon/ion/impl/bin/PresenceBitmap.kt

```diff
@@ -0,0 +1,220 @@
+ *    - Since all the parameter cardinalities (other than `!`) use the same presence bit semantics, the writer doesn't
+ *      need to inspect the signature to figure out what bits to put in our presence bits buffer.
+ *    - Because we have dummy bits for `!` parameters, [PresenceBits] can present an API that corresponds 1:1 with
+ *      parameters, so we don't need to separately keep track of a presence bit index and the parameter count.
+ *  - Why longs instead of an array?
+ *    - An array would add another level of indirection
+ *    - An array would require a loop in order to reset all the bytes to zero.
+ *  - Why only 128 parameters?
+ *    - Until proven otherwise, we should not assume that an arbitrarily large number of parameters MUST be supported.
+ *    - The number of parameters could be increased (within limits). It seems reasonable to try to keep this class small
+ *      enough to fit in a single cache line for a modern system—typically 64 bytes.
+ *
+ * TODO: Consider whether we can "compile" a specific function that can read the presence bits when we compile a macro.
+ *       That _might_ be more efficient than this approach.
+ */
+internal class PresenceBitmap {
+
+    companion object {
+        const val VOID = 0b00L
+        const val EXPRESSION = 0b01L
+        const val GROUP = 0b10L
+        const val RESERVED = 0b11L
+
+        private const val TWO_BIT_MASK = 0b11L
+        private const val PRESENCE_BITS_SIZE_THRESHOLD = 2
+        private const val PB_SLOTS_PER_BYTE = 4
+        private const val PB_SLOTS_PER_LONG = 32
+        private const val PB_BITS_PER_SLOT = 2
+
+        const val MAX_SUPPORTED_PARAMETERS = PB_SLOTS_PER_LONG * 4
+    }
+
+    private var signature: List<Parameter> = emptyList()
+
+    /** The number of parameters for which presence bits must be written. */
+    private var size: Int = 0
+
+    /** The total number of parameters in the macro signature */
+    val totalParameterCount: Int
+        get() = signature.size
+
+    /** The first 32 presence bits slots */
+    private var a: Long = 0
+    /** The second 32 presence bits slots */
+    private var b: Long = 0
+    /** The third 32 presence bits slots */
+    private var c: Long = 0
+    /** The fourth 32 presence bits slots */
+    private var d: Long = 0
+
+    /** The number of bytes required to encode this [PresenceBitmap] */
+    val byteSize: Int
+        get() = size divideByRoundingUp PB_SLOTS_PER_BYTE
+
+    /** Resets this [PresenceBitmap] for the given [macro]. */
+    fun initialize(signature: List<Parameter>) {
+        a = 0
+        b = 0
+        c = 0
+        d = 0
+        size = calculateNumPresenceBits(signature)
+        this.signature = signature
+    }
+
+    /**
+     * Checks that all presence bits are valid for their corresponding parameters.
+     * Throws [IonException] if any are not.
+     */
+    fun validate() {
+        signature.forEachIndexed { i, it ->
+            val presenceValue = get(i)
```

Comment:
So, after seeing https://github.com/amazon-ion/ion-tests/pull/120, I'm not sure whether this method is even useful anymore. Do we want to check how the argument is encoded, or should we check the actual number of expressions produced/provided by the argument?

Reply:
This applies here and for performance of this class in general.

I tried adding an iterator, and then it occurred to me that an Iterator can only hold boxed values, so I looked at the byte code to see what was going on.

For `get`:
```
   L2
   FRAME SAME
    ILOAD 1
    BIPUSH 32
    IDIV
    TABLESWITCH
      0: L3
      1: L4
      2: L5
      3: L6
      default: L7
   L3
   FRAME SAME
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap.a : J
    GOTO L8
   L4
   FRAME SAME
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap.b : J
    GOTO L8
   L5
   FRAME SAME
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap.c : J
    GOTO L8
   L6
   FRAME SAME
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap.d : J
    GOTO L8

<I've omitted the unreachable default case>

   L8
   FRAME SAME1 J
    LSTORE 2
   L9
    ILOAD 1
    BIPUSH 32
    IREM
    ICONST_2
    IMUL
    ISTORE 4
   L10
    LLOAD 2
    ILOAD 4
    LSHR
    LDC 3
    LAND
    LRETURN
```

It looks like a lot, but all of these are very cheap operations. If all macros have 32 parameters or fewer, then the JIT can completely optimize away the `TABLESWITCH`, making this a completely branchless set of local load/store operations and math operations.

Using an iterator, on the other hand, requires two `INVOKEVIRTUAL` just to get the next `long` from the iterator.
```
L8
    ALOAD 2
    INVOKEVIRTUAL com/amazon/ion/impl/bin/PresenceBitmap$Iter.next ()Ljava/lang/Long;
    INVOKEVIRTUAL java/lang/Long.longValue ()J
    LSTORE 7
```
And in the iterator, in order to produce the next value:
```
L0
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.i : I
    BIPUSH 32
    IREM
    IFNE L1
   L2
    ALOAD 0
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.i : I
    BIPUSH 32
    IDIV
    TABLESWITCH
      0: L3
      1: L4
      2: L5
      3: L6
      default: L7
   L3
   FRAME SAME1 com/amazon/ion/impl/bin/PresenceBitmap$Iter
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.a : J
    GOTO L8
   L4
   FRAME SAME1 com/amazon/ion/impl/bin/PresenceBitmap$Iter
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.b : J
    GOTO L8
   L5
   FRAME SAME1 com/amazon/ion/impl/bin/PresenceBitmap$Iter
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.c : J
    GOTO L8
   L6
   FRAME SAME1 com/amazon/ion/impl/bin/PresenceBitmap$Iter
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.d : J
    GOTO L8

<unreachable default case omitted>

   L8
   FRAME FULL [com/amazon/ion/impl/bin/PresenceBitmap$Iter] [com/amazon/ion/impl/bin/PresenceBitmap$Iter J]
    PUTFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.currentLong : J
   L1
   FRAME SAME
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.currentLong : J
    LDC 3
    LAND
    LSTORE 1
   L9
    LINENUMBER 273 L9
    ALOAD 0
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.currentLong : J
    ICONST_2
    LSHR
    PUTFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.currentLong : J
   L10
    ALOAD 0
    GETFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.i : I
    ISTORE 3
    ALOAD 0
    ILOAD 3
    ICONST_1
    IADD
    PUTFIELD com/amazon/ion/impl/bin/PresenceBitmap$Iter.i : I
   L11
    LLOAD 1
    INVOKESTATIC java/lang/Long.valueOf (J)Ljava/lang/Long;
    ARETURN
```

There's a lot of similarity. The main thing, though, is that we're having to store the iterator state in the fields of the Iterator instead of the local stack, and there's an extra branch. I could get rid of some of that extra branching and extra state, but then `next()` would just be a wrapper around `get(i)` (but still with extra boxing of the `long` values).

I'm going to leave this as is, and we can revisit it if it does prove to be a bottleneck.

Reply:
Well, I forgot that you can create a private iterator without implementing `Iterator` so that it can return `long` instead of `Long`. That saves us from the auto-boxing, but it still doesn't seem like it's saving much in the way of computation, and it does add the overhead of an extra class.